_posts/2022-07-18-infinispan-14-opentelemetry-tracing.adoc

```diff
@@ -0,0 +1,164 @@
+
+[source, properties]
+----
+infinispan.tracing.enabled=true
+otel.service.name=infinispan-server-service
+otel.exporter.otlp.endpoint=http://localhost:4317
+otel.metrics.exporter=none
+----
+
+The first property is Infinispan specific and enables the tracing capability of the Infinispan server.
+The further properties belong to the https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/README.md[OpenTelemetry SDK Autoconfigure project]
+that Infinispan uses to configure the tracing exporter.
+In this case OTLP gRPC exporter protocol is used and the server runs in the same machine of the Infinispan server.
+
+Starting the server with these parameters:
+[source, shell script]
+----
+export JAVA_OPTS="-Dinfinispan.tracing.enabled=true -Dotel.service.name=infinispan-server-service -Dotel.exporter.otlp.endpoint=http://localhost:4317 -Dotel.metrics.exporter=none"
+./server.sh
+----
+
+The following log is produced when the server starts:
+[source, bash]
+----
+(ForkJoinPool.commonPool-worker-2) [org.infinispan.server.core.telemetry.TelemetryServiceFactory] ISPN000952: OpenTelemetry instance loaded: OpenTelemetrySdk{...
+----
+
+To indicate the fact that the OpenTelemetrySdk is correctly configured.
+
+== Tracing from a HotRod client application
+
+Any OpenTelemetry tracing context present on HotRot client applications will be automatically propagated by the new Hot Rod v4 client to the server tracing context.
+
+For instance for a client defining some tracing spans containing cache operations, such as the following:
+
+[source,java]
+----
+public class MyRestClient {
+
+    public void putSomeValues(RemoteCache cache) {
+        Span span = tracer.spanBuilder("sub-bulk-1").setSpanKind(SpanKind.CLIENT).startSpan();
+        // put the span into the current Context
+        try (Scope scope = span.makeCurrent()) {
+            cache.put(1, "A");
+            cache.put(2, "B");
+            cache.put(3, "C");
+        } catch (Throwable throwable) {
+            span.setStatus(StatusCode.ERROR, "Something bad happened!");
+            span.recordException(throwable);
+            throw throwable;
+        } finally {
+            span.end(); // Cannot set a span after this call
+        }
+    }
+}
+----
+
+the client span `sub-bulk-1` will be correlated to any related server spans, in this case the three put operations.
+
+Opening the Jaeger console, we can see that client and server spans are correctly aggregated:
+
+[caption="Jaeger console",link=/assets/images/blog/tracing-hotrod-client.png]
+image::/assets/images/blog/tracing-hotrod-client.png[Tracing output]
+
+You can find a complete application example here: https://github.com/fax4ever/infinispan-play/tree/main/tracing-hotrod-client
+
+== Tracing from a REST client application
+
+The same result can be achieved from a Rest client, putting manually in the HTTP headers requests the current tracing context
```

Comment:
You can achieve the same with a REST client by putting manually in the HTTP headers requests to provide information about the current tracing context using a standard a OpenTelemetry instance of `W3CTraceContextPropagator`. ?